src/WebJobs.Extensions.Http/ClaimsPrincipalBindingProvider.cs

```diff
@@ -0,0 +1,116 @@
+using System.Linq;
+using System.Net.Http;
+using System.Reflection;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Azure.WebJobs.Host.Bindings;
+using Microsoft.Azure.WebJobs.Host.Protocols;
+using static Microsoft.Azure.WebJobs.Extensions.Http.HttpTriggerAttributeBindingProvider.HttpTriggerBinding;
+
+namespace Microsoft.Azure.WebJobs.Extensions.Http
+{
+    /// <summary>
+    /// This provider provides a binding to Type <see cref="ClaimsPrincipal"/>.
+    /// </summary>
+    internal class ClaimsPrincipalBindingProvider : IBindingProvider
+    {
+        private static readonly Task<IBinding> NullBinding = Task.FromResult<IBinding>(null);
+
+        private static readonly IReadOnlyCollection<string> PotentialIdentityHeaders = new HashSet<string>
+        {
+            "x-ms-client-principal", 
+            "x-ms-functions-key-identity"
+        };
+
+        public Task<IBinding> TryCreateAsync(BindingProviderContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
+            }
+
+            if (context.Parameter.ParameterType != typeof(ClaimsPrincipal))
+            {
+                return NullBinding;
+            }
+
+            return Task.FromResult<IBinding>(new ClaimsIdentityBinding(context.Parameter));
+        }
+
+        private class ClaimsIdentityBinding : IBinding
+        {
+            private readonly ParameterInfo _parameter;
+
+            public ClaimsIdentityBinding(ParameterInfo parameter)
+            {
+                _parameter = parameter;
+            }
+
+            public bool FromAttribute
+            {
+                get { return false; }
+            }
+
+            public Task<IValueProvider> BindAsync(BindingContext context)
+            {
+                if (context == null)
+                {
+                    throw new ArgumentNullException("context");
+                }
+
+                IReadOnlyDictionary<string, object> bindingData = context.BindingData;
+                ClaimsPrincipal principal = GetClaimsPrincipalFromBindingData(bindingData);
+
+                var valueProvider = new SimpleValueProvider(typeof(ClaimsPrincipal), principal, principal?.Identity?.Name);
+                return Task.FromResult<IValueProvider>(valueProvider);
+            }
+
+            private static ClaimsPrincipal GetClaimsPrincipalFromBindingData(IReadOnlyDictionary<string, object> bindingData)
+            {
+                object headers;
+                if (bindingData.TryGetValue(HttpTriggerAttributeBindingProvider.HttpHeadersKey, out headers))
+                {
+                    List<ClaimsIdentity> identities = (headers as IDictionary<string, string>)
+                        .Where(header => PotentialIdentityHeaders.Any(idHeader => string.Equals(idHeader, header.Key, StringComparison.OrdinalIgnoreCase)))
```

Comment:
>PotentialIdentityHeaders [](start = 41, length = 24)

Why isn't this simply `PotentialIdentityHeaders.Contains(header.Key)? It seems that would be more efficient. You can achieve case-insensitivity by using a case-insensitive HashSet<string> above.

Reply:
I had no idea that existed. That was exactly what I was going for!

---
In reply to: [181944504](https://github.com/Azure/azure-webjobs-sdk-extensions/pull/404#discussion_r181944504) [](ancestors = 181944504)